SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/EntityComponentTestUtils.h

```diff
@@ -0,0 +1,180 @@
+	Schema_SerializeToBuffer(Lhs, LhsBuffer.Get(), Length);
+	Schema_SerializeToBuffer(Rhs, RhsBuffer.Get(), Length);
+	return FMemory::Memcmp(LhsBuffer.Get(), RhsBuffer.Get(), Length) == 0;
+}
+
+/** Returns true if Lhs and Rhs have the same component ID and state. */
+inline bool CompareComponentData(const ComponentData& Lhs, const ComponentData& Rhs)
+{
+	if (Lhs.GetComponentId() != Rhs.GetComponentId())
+	{
+		return false;
+	}
+	return CompareSchemaObjects(Lhs.GetFields(), Rhs.GetFields());
+}
+
+/** Returns true if Lhs and Rhs have the same component ID and events. */
+inline bool CompareComponentUpdateEvents(const ComponentUpdate& Lhs, const ComponentUpdate& Rhs)
+{
+	if (Lhs.GetComponentId() != Rhs.GetComponentId())
+	{
+		return false;
+	}
+	return CompareSchemaObjects(Lhs.GetEvents(), Rhs.GetEvents());
+}
+
+/** Returns true if Lhs and Rhs have the same component ID and state. */
+inline bool CompareComponentUpdates(const ComponentUpdate& Lhs, const ComponentUpdate& Rhs)
+{
+	if (Lhs.GetComponentId() != Rhs.GetComponentId())
+	{
+		return false;
+	}
+	return CompareSchemaObjects(Lhs.GetFields(), Rhs.GetFields()) &&
+		CompareSchemaObjects(Lhs.GetEvents(), Rhs.GetEvents());
+}
+
+/** Returns true if Lhs and Rhs have the same entity ID, component ID, and state. */
+inline bool CompareEntityComponentData(const EntityComponentData& Lhs,
+	const EntityComponentData& Rhs)
+{
+	if (Lhs.EntityId != Rhs.EntityId)
+	{
+		return false;
+	}
+	return CompareComponentData(Lhs.Data, Rhs.Data);
+}
+
+/** Returns true if Lhs and Rhs have the same entity ID, component ID, and events. */
+inline bool CompareEntityComponentUpdateEvents(const EntityComponentUpdate& Lhs,
+	const EntityComponentUpdate& Rhs)
+{
+	if (Lhs.EntityId != Rhs.EntityId)
+	{
+		return false;
+	}
+	return CompareComponentUpdateEvents(Lhs.Update, Rhs.Update);
+}
+
+/** Returns true if Lhs and Rhs have the same entity ID, component ID, state, and events. */
+inline bool CompareEntityComponentUpdates(const EntityComponentUpdate& Lhs,
+	const EntityComponentUpdate& Rhs)
+{
+	if (Lhs.EntityId != Rhs.EntityId)
+	{
+		return false;
+	}
+	return CompareComponentUpdates(Lhs.Update, Rhs.Update);
+}
+
+/** Returns true if Lhs and Rhs have the same ID, component ID, data, state and events. */
+inline bool CompareEntityComponentCompleteUpdates(const EntityComponentCompleteUpdate& Lhs,
+	const EntityComponentCompleteUpdate& Rhs)
+{
+	if (Lhs.EntityId != Rhs.EntityId)
+	{
+		return false;
+	}
+	return CompareComponentData(Lhs.CompleteUpdate, Rhs.CompleteUpdate) && CompareComponentUpdateEvents(Lhs.Events, Rhs.Events);
+}
+
+inline bool CompareEntityComponentId(const EntityComponentId& Lhs, const EntityComponentId& Rhs)
+{
+	return Lhs == Rhs;
+}
+
+template<typename T>
+using ComparisonFunction = bool(*)(const T& Lhs, const T& Rhs);
+
+template<typename T>
+bool AreEquivalent(const TArray<T>& Lhs, const TArray<T>& Rhs, ComparisonFunction<T> Compare)
```

Comment:
Is this meant to be IsPerumtation?
Why not just make Compare a template type and pass by universal reference. This is probably fine it's just more prescriptive of the argument than it needs to be.
If we do this we could also use std::is_permutation with the underlying pointers from TArray.

```
template<typename T, typename Predicate>
bool AreEquivalent(const TArray<T>& Lhs, const TArray<T>& Rhs, Predicate&& Comparison)
{
    return std::is_permutation(Lhs.Data(), Lhs.Data() + Lhs.Num(), rhs.Data(), std::forward<Predicate>(Comparison));
}
```

Reply:
Everything except using `std` function sounds good. I can also add `std` function call as long as we agree that it doesn't contradict our coding standard?

Reply:
Made Compare a template type.

Reply:
Added is_permutation